src/allmydata/storage_client.py

```diff
@@ -1198,16 +1188,16 @@ def try_to_connect(self):
 
     @async_to_deferred
     async def _connect(self):
```

Comment:
It seems like `_connect` is not re-entrant (either before or after this change).  Possibly it doesn't need to be since it is used with `LoopingCall` but it seems like a bit of a trap to me.

The unsafety comes from the handling `_connecting_deferred` which could be written and then written again if a second call is made to `_connect` before a first completes.

Maybe this is just worth a note in a docstring.

Reply:
It's also used with try_to_connect() so it could happen. I'll see if I can fix it.

Reply:
I added an assert that reentrancy isn't happening, and now sometimes tests fail :cry: